airflow/utils/sla.py

```diff
@@ -0,0 +1,371 @@
+    Return a string representation of the task instance.
+    """
+    return "{dag_id}.{task_id}[{exc_date}]".format(
+        dag_id=ti.dag_id,
+        task_id=ti.task_id,
+        exc_date=ti.execution_date
+    )
+
+
+def get_sla_miss_subject(miss_type, ti):
+    """
+    Return a consistent subject line for SLA miss emails.
+    """
+    return "[airflow] [SLA] {miss_type} on {task_instance}".format(
+        miss_type=miss_type,
+        task_instance=describe_task_instance(ti)
+    )
+
+
+def get_subscribers(tasks):
+    """
+    Return a list of unique emails from a list of tasks.
+    """
+    def _yield_subscribers(tasks):
+        for t in tasks:
+            if t.email:
+                if isinstance(t.email, basestring):
+                    yield t.email
+                else:
+                    for e in t.email:
+                        yield e
+    return list(set(_yield_subscribers(tasks)))
+
+
+@provide_session
+def get_blocked_task_instances(task_instance, session=None):
+    """
+    Given a task instance, return task instances that may currently
+    be blocked by it.
+    """
+    dag = task_instance.task.dag
+
+    TI = airflow.models.TaskInstance
+    downstream_tasks = task_instance.get_flat_relatives(upstream=False)
+
+    # The intent is to capture states that indicate that work was never started
+    # on a task, presumably because this task not achieving its SLA prevented
+    # the downstream task from ever successfully starting.
+
+    # It is possible for an upstream task to cause a downstream task to *fail*,
+    # like if it never produces a required artifact. But in a well-behaved DAG
+    # where dependencies are encoded properly, this shouldn't happen.
+    blocked_states = (
+        State.UPSTREAM_FAILED,
+        State.SCHEDULED,
+        State.QUEUED,
+        State.NONE,
+    )
+
+    qry = (
+        session
+        .query(TI)
+        .filter(TI.dag_id == dag.dag_id)
+        .filter(TI.task_id.in_(list(t.task_id for t in downstream_tasks)))
+        .filter(TI.execution_date == task_instance.execution_date)
+        .filter(TI.state.in_(blocked_states))
+        .all()
+    )
+    return qry
+
+
+def send_task_duration_exceeded_email(context):
+    ti = context["ti"]
+    target_time = ti.task.expected_duration
+    blocked = get_blocked_task_instances(ti)
+
+    email_to = get_subscribers(blocked)
+    email_subject = get_sla_miss_subject("Exceeded duration", ti)
+    email_body = ""
+    "<pre><code>{task_string}</pre></code> missed an SLA: duration "
+    "exceeded <pre><code>{target_time}</pre></code>.\n\n"
+
+    "View Task Details: {ti_url}\n\n"
+
+    "This may be blocking the following downstream tasks:\n"
+
+    "<pre><code>{blocked_tasks}\n{art}</pre></code>".format(
+        task_string=describe_task_instance(ti),
+        target_time=target_time,
+        blocked_tasks="\n".join(describe_task_instance(d) for d in blocked),
+        ti_url=ti.details_url,
+        art=asciiart.snail)
```

Comment:
Is the snail art really necessary?

Reply:
I think it better communicates that something is running slow :)